File: mobile/lib/shared/models/int_math.dart
/// Overflow-safe integer arithmetic for financial values.
///
/// Dart port of `App\Modules\Shared\Support\IntMath`. Where the PHP version
/// routes intermediate products through bcmath, this version routes them
/// through [BigInt] — the reason is the same in both languages: the product
/// `fine_weight_mg * price_per_gram_rial` for a large lot is ~10^16, which
/// still fits in 64 bits, but only just, and there is no margin left for a
/// second multiplication before an overflow silently produces a wrong number.
///
/// See docs/11-appendix/01-formulas.md and ADR-002.
library;

final class IntMath {
  const IntMath._();

  /// Largest value representable by a Dart native `int`.
  static final BigInt maxInt = BigInt.parse('9223372036854775807');

  /// Smallest value representable by a Dart native `int`.
  static final BigInt minInt = BigInt.parse('-9223372036854775808');

  /// `floor(a * b / c)` with an arbitrary-precision intermediate product.
  ///
  /// Throws [ArithmeticOverflowError] if the result does not fit in an `int`,
  /// and [ArgumentError] if [c] is zero.
  static int mulDivFloor(int a, int b, int c) {
    if (c == 0) {
      throw ArgumentError.value(c, 'c', 'Division by zero in mulDivFloor');
    }

    final product = BigInt.from(a) * BigInt.from(b);

    return _toInt(_divFloor(product, BigInt.from(c)));
  }

  /// `ceil(a * b / c)` with an arbitrary-precision intermediate product.
  ///
  /// Throws [ArithmeticOverflowError] if the result does not fit in an `int`,
  /// and [ArgumentError] if [c] is zero.
  static int mulDivCeil(int a, int b, int c) {
    if (c == 0) {
      throw ArgumentError.value(c, 'c', 'Division by zero in mulDivCeil');
    }

    final product = BigInt.from(a) * BigInt.from(b);

    return _toInt(_divCeil(product, BigInt.from(c)));
  }

  /// The remainder discarded by [mulDivFloor]: `a*b - floor(a*b/c)*c`.
  ///
  /// The server routes this dust into the `ROUNDING_DIFFERENCE` system account
  /// so that mass conservation holds. The client only ever displays it.
  static int mulDivRemainder(int a, int b, int c) {
    if (c == 0) {
      throw ArgumentError.value(c, 'c', 'Division by zero in mulDivRemainder');
    }

    final divisor = BigInt.from(c);
    final product = BigInt.from(a) * BigInt.from(b);
    final quotient = _divFloor(product, divisor);

    return _toInt(product - quotient * divisor);
  }

  /// Addition with overflow detection.
  static int add(int a, int b) => _toInt(BigInt.from(a) + BigInt.from(b));

  /// Subtraction with overflow detection.
  static int subtract(int a, int b) => _toInt(BigInt.from(a) - BigInt.from(b));

  /// Multiplication with overflow detection.
  static int multiply(int a, int b) => _toInt(BigInt.from(a) * BigInt.from(b));

  /// Sum of [values] with overflow detection.
  static int sum(Iterable<int> values) {
    var total = BigInt.zero;
    for (final value in values) {
      total += BigInt.from(value);
    }

    return _toInt(total);
  }

  /// BigInt `~/` truncates toward zero. Ledger amounts are signed, so a true
  /// floor is required: `-3 ~/ 2` is `-1`, but `floor(-3/2)` is `-2`.
  static BigInt _divFloor(BigInt numerator, BigInt denominator) {
    final quotient = numerator ~/ denominator;
    final remainder = numerator - quotient * denominator;

    if (remainder != BigInt.zero && _signsDiffer(numerator, denominator)) {
      return quotient - BigInt.one;
    }

    return quotient;
  }

  static BigInt _divCeil(BigInt numerator, BigInt denominator) {
    final quotient = numerator ~/ denominator;
    final remainder = numerator - quotient * denominator;

    if (remainder != BigInt.zero && !_signsDiffer(numerator, denominator)) {
      return quotient + BigInt.one;
    }

    return quotient;
  }

  static bool _signsDiffer(BigInt a, BigInt b) => a.isNegative != b.isNegative;

  static int _toInt(BigInt value) {
    if (value > maxInt || value < minInt) {
      throw ArithmeticOverflowError(value.toString());
    }

    return value.toInt();
  }
}

/// Thrown when a financial computation produces a value outside the 64-bit
/// integer range. This is never recoverable in the UI: it means the inputs are
/// nonsensical, so it surfaces as a hard error rather than a formatted zero.
class ArithmeticOverflowError extends Error {
  ArithmeticOverflowError(this.value);

  final String value;

  @override
  String toString() =>
      'ArithmeticOverflowError: $value does not fit in a 64-bit integer';
}

File: app/Modules/Shared/Support/IntMath.php
<?php

declare(strict_types=1);

namespace App\Modules\Shared\Support;

use App\Modules\Shared\Exceptions\ArithmeticOverflowException;
use DivisionByZeroError;

/**
 * Overflow-safe integer arithmetic for financial values.
 *
 * Every financial quantity in this system is a 64-bit integer (milligrams for
 * weight, rial for money). Intermediate products can exceed PHP_INT_MAX, so all
 * multiply-then-divide operations route through bcmath and are checked before
 * being handed back as int.
 *
 * See docs/11-appendix/01-formulas.md and ADR-002.
 */
final class IntMath
{
    private function __construct() {}

    /**
     * floor(a * b / c) with an arbitrary-precision intermediate product.
     *
     * @throws ArithmeticOverflowException when the result does not fit in a 64-bit int
     */
    public static function mulDivFloor(int $a, int $b, int $c): int
    {
        if ($c === 0) {
            throw new DivisionByZeroError('Division by zero in mulDivFloor');
        }

        $product = bcmul((string) $a, (string) $b, 0);
        $result = self::bcDivFloor($product, (string) $c);

        return self::toInt($result);
    }

    /**
     * ceil(a * b / c) with an arbitrary-precision intermediate product.
     *
     * @throws ArithmeticOverflowException when the result does not fit in a 64-bit int
     */
    public static function mulDivCeil(int $a, int $b, int $c): int
    {
        if ($c === 0) {
            throw new DivisionByZeroError('Division by zero in mulDivCeil');
        }

        $product = bcmul((string) $a, (string) $b, 0);
        $result = self::bcDivCeil($product, (string) $c);

        return self::toInt($result);
    }

    /**
     * The remainder discarded by mulDivFloor, expressed in units of the divisor's
     * numerator scale: a*b - floor(a*b/c)*c. Used to route rounding dust into the
     * ROUNDING_DIFFERENCE system account so mass conservation still holds.
     */
    public static function mulDivRemainder(int $a, int $b, int $c): int
    {
        if ($c === 0) {
            throw new DivisionByZeroError('Division by zero in mulDivRemainder');
        }

        $product = bcmul((string) $a, (string) $b, 0);
        $quotient = self::bcDivFloor($product, (string) $c);
        $remainder = bcsub($product, bcmul($quotient, (string) $c, 0), 0);

        return self::toInt($remainder);
    }

    /** Addition with overflow detection. */
    public static function add(int $a, int $b): int
    {
        return self::toInt(bcadd((string) $a, (string) $b, 0));
    }

    /** Subtraction with overflow detection. */
    public static function sub(int $a, int $b): int
    {
        return self::toInt(bcsub((string) $a, (string) $b, 0));
    }

    /** Multiplication with overflow detection. */
    public static function mul(int $a, int $b): int
    {
        return self::toInt(bcmul((string) $a, (string) $b, 0));
    }

    /**
     * Sum of a list with overflow detection.
     *
     * @param  iterable<int>  $values
     */
    public static function sum(iterable $values): int
    {
        $total = '0';
        foreach ($values as $value) {
            $total = bcadd($total, (string) $value, 0);
        }

        return self::toInt($total);
    }

    /**
     * bcdiv truncates toward zero; we need a true floor for negative numerators
     * because ledger entries are signed.
     */
    private static function bcDivFloor(string $numerator, string $denominator): string
    {
        $quotient = bcdiv($numerator, $denominator, 0);

        // Truncation equals floor unless there is a remainder and signs differ.
        $remainder = bcsub($numerator, bcmul($quotient, $denominator, 0), 0);

        if (bccomp($remainder, '0', 0) !== 0 && self::signsDiffer($numerator, $denominator)) {
            $quotient = bcsub($quotient, '1', 0);
        }

        return $quotient;
    }

    private static function bcDivCeil(string $numerator, string $denominator): string
    {
        $quotient = bcdiv($numerator, $denominator, 0);
        $remainder = bcsub($numerator, bcmul($quotient, $denominator, 0), 0);

        if (bccomp($remainder, '0', 0) !== 0 && ! self::signsDiffer($numerator, $denominator)) {
            $quotient = bcadd($quotient, '1', 0);
        }

        return $quotient;
    }

    private static function signsDiffer(string $a, string $b): bool
    {
        $aNegative = bccomp($a, '0', 0) < 0;
        $bNegative = bccomp($b, '0', 0) < 0;

        return $aNegative !== $bNegative;
    }

    /** @throws ArithmeticOverflowException */
    private static function toInt(string $value): int
    {
        if (bccomp($value, (string) PHP_INT_MAX, 0) > 0 || bccomp($value, (string) PHP_INT_MIN, 0) < 0) {
            throw new ArithmeticOverflowException($value);
        }

        return (int) $value;
    }
}
